Open file with shifted layout

achaikou opened the pull request:
With this PR we can open files where trace fields are located not at their default position.

We cannot yet open our available 2.1 file, cannot correctly read fields and do not support non-standard datatypes.
Everythin related to that would happen in the next PRs.

achaikou committed:
Test layout with non-standard positions

achaikou committed:
Relocate D8 parsing tables

Some other functions would need the content.

achaikou committed:
Return traceheader mappings to python

Return final traceheader layout, regardless of it being taken from xml,
default specification, or overwritten by the user.

It might be that this could would look nicer if we ever switch to
pybind11.

achaikou committed:
Assign names to default traceheader mappings

When xml is not present, we use default maps from C, which for the
moment have their name set to NULL.

But users do not care where information came from and likely would want
to know the names as well, so they must be assigned.

---
As revision 2.1 introduced strings of unknown length and other
unknown-size features, dealing with this in C became more complicated.

Current solution where implementation happens in C++ bindings part
instead of C core is not optimal, but it works good enough in python
which is a main target user language.

There is no good vision yet what optimal solution would be.

ajaust commented on python/segyio/utils.py:
```diff
@@ -146,6 +146,25 @@ def __repr__(self):
+    def __init__(self, entries):
+        self.entries = list(entries)
+        self.by_byte = {entry.byte: entry for entry in self.entries}
+        self.by_name = {entry.name: entry for entry in self.entries}
```
Should the members be marked as private, i.e, prefix `entries, `by_byte`, and `by_name` with `_`?

ajaust commented on python/segyio/utils.py:
```diff
@@ -146,6 +146,25 @@ def __repr__(self):
+        return self.by_byte.get(byte)
+
+    def get_by_name(self, name):
+        return self.by_name.get(name)
```
Depending on how we want to access/work with the members of the class, we could drop the `get_` methods and directly access the members similar to the approach in `TraceHeaderLayoutEntry`.

ajaust commented on python/segyio/segyio.cpp:
```diff
@@ -2772,6 +2780,127 @@ extern "C" int segy_parse_layout_xml(
+            entry_type = entry_type_map[i].spec_entry_type;
+            break;
+        }
+    }
```
If this is actually too expensive, search for every entry type, I guess the simplest is to set up some lookup table/method to avoid the search.

ajaust commented on python/segyio/open.py:
```diff
@@ -80,12 +80,16 @@ def open(filename, mode="r", iline = 189,
+        Overrides crossline field offset in the trace headers.  If no value is
+        set, value defined in xml mapping or standard-defined 193 would be used.
+        By setting this value user takes over responsibility of assuring parsing
+        can be done correctly.
```
Just a reminder that if we update the description for the `iline` parameter, we should update this accordingly.

ajaust commented on test-data/README.md:
```diff
@@ -97,5 +97,6 @@ Files are created manually.
 | mapping-invalid-root.sgy       | Root does not define standard header layout.                  |
 | mapping-invalid-value.sgy      | Last standard header entry byte is out of range.              |
 | mapping-multiple-stanzas.sgy   | Other stanzas are present before and after layout stanza.     |
+| mapping-shifted.sgy            | Default mapping with all offsets shifted by 1 byte.           |
```
Just for clarification. The mapping is embedded inside as XML and therefore we do not have to specify the offsets when opening the files inside the tests. Is this correct?

ajaust reviewed (commented):
Overall, I think the changes look good. I think my main concern/confusion is when working with the dictionary of layouts or a specific layout and the corresponding entries. I am not sure if we are able to make the code simpler to read if we would rename some variable. I made some suggestions, but I am not sure if there is a better solution. I am curious about your thoughts. 🙂

achaikou commented on python/segyio/segyio.cpp:
```diff
@@ -507,6 +507,12 @@ int set_traceheader_mappings(
 int extract_layout_stanza( segyfd* self, std::vector<char>& layout_stanza_data );
 
+/**
+ * Returns the dictionary {header_name: [TraceHeaderLayoutEntry] of traceheader
```
Fix the docs

achaikou reviewed (commented):
Thanks for the review!
Something is updated, something will be updated after we agree upon it :slightly_smiling_face:

ajaust reviewed (approved):
The changes look good. I don't have anything in mind that must be changed. 😄 

I am just curious to see the code using `layout`/`layouts` part to get an impression whether this looks better, *but only* if it is not too much work.

ajaust commented on python/test/segyio_c.py:
```diff
@@ -182,7 +182,7 @@ def test_custom_iline_xline():
-    layout = f.traceheader_layout()
+    layout = f.traceheader_layouts()
 
     standard_header_layout = layout["SEG00000"]
```
If we keep `layouts`, then maybe also update the temporary variable:
```suggestion
    layouts = f.traceheader_layouts()

    standard_header_layout = layouts["SEG00000"]

achaikou replied:
Done, but as I rebased, it is not a fixup :)

ajaust commented on python/segyio/segyio.cpp:
```diff
@@ -507,6 +507,12 @@ int set_traceheader_mappings(
+ * Returns the dictionary {header_name: [TraceHeaderLayoutEntry] of traceheader
+ * mappings.}
+ */
+PyObject* traceheader_layout( segyfd* self );
```
Great. I think I prefer the new version using `layouts`. 🙂

ajaust reviewed (approved):
Great! Thanks for addressing this so quickly. It looks good from my side. I had just one minor suggestion in the newly added fixups.

achaikou committed:
Get rid of iline/xline defaults in python

Now iline/xline can be defined in xml, but user-assigned iline/xline
override them.
Thus we should not supply anything as default values for user-assigned
iline/xline and instead retrieve actual values from mappings.

achaikou committed:
Open segy_field_forall for all types

segy_field_forall opens for support of all types, not only int32.

- Full support for float types is not implemented yet.
- Works only on standard trace header. Design to support other header is
not decided upon yet.
- "indicies" function still supports only int32 inline-xline.
- scale6 type is not supported as it is unclear yet how to deal with it.

achaikou commented on python/segyio/segy.py:
```diff
@@ -27,14 +27,15 @@ def __init__(
-        self._il = iline
-        self._xl = xline
+
+        self.traceheader_layouts = fd.traceheader_layouts()
```
Maybe _this_ should be private though.
It is likely not the final way we want users to access the data, so maybe we should keep it kind-of-hidden.

yngve793 commented on python/segyio/segyio.cpp:
```diff
@@ -2733,6 +2751,122 @@ extern "C" int segy_parse_layout_xml(
+}
+
+PyObject* traceheader_layouts( segyfd* self ) {
+    PyObject* module = PyImport_ImportModule( "segyio.utils" );
```
The word `module` is flagged as a keyword in Github and VScode. Consider changing the variable name slightly.

yngve793 replied:
Nice, fix.

yngve793 reviewed (approved):
The PR is in my opinion fine. Added some minor comments and suggestions. None of them are critical or principally important.  

1) Setting names to NULL in C.  We have discussed this in multiple iterations and the limitations in C are what they are. All in all I consider the final solution to be a good choice given the available options. 

2) `segy_field_forall` not supporting float. This is fine for now, but do we have a clear idea on how to incorporate this without breaking to much of the existing code structure? If we have a rough plan then that is great. If not it would make sense to rank importance versus implications.

achaikou reviewed (commented):
Thanks for the review!!

2. This does not support floats yet because floats are not added to all switch-cases. So I think it should be dealt with in the next PR without any additional problems.

yngve793 commented on python/segyio/segyio.cpp:
```diff
@@ -2379,7 +2379,16 @@ int overwrite_inline_xline(
         uint8_t il = static_cast<uint8_t>( iline );
         mapping.name_to_offset[SEGY_TR_INLINE] = il;
-        mapping.offset_to_entry_definition[il - 1] = { SEGY_ENTRY_TYPE_INT4, false, NULL };
+
```
Thanks, very nice.
